s5a/__init__.py

```diff
@@ -6,80 +6,72 @@
+    """
 
-    def __repr__(self):
-        return (
-            f'longitude={self.longitude} latitude={self.latitude} '
-            f'value={self.value} timestamp={self.timestamp} '
-            f'quality={self.quality}')
+    # read in data
+    with netCDF4.Dataset(ncfile, 'r') as f:
+        variables = f.groups['PRODUCT'].variables
+        data = variables['carbonmonoxide_total_column'][:][0]
+        longitude = variables['longitude'][:][0]
+        latitude = variables['latitude'][:][0]
+        quality = variables['qa_value'][:][0]
+        deltatime = variables['delta_time'][:][0]
+        meta_data = f.__dict__
 
+    # get some metadata
+    # use mask from MaskedArray to filter values
+    mask = numpy.logical_not(data.mask)
+    n_lines = data.shape[0]  # number of scan lines
+    pixel_per_line = data.shape[1]  # number of pixels per line
+    time_reference = meta_data['time_reference_seconds_since_1970']
 
-class Scan():
-    """Object to hold arrays from an nc file.
-    """
-    def __init__(self, filepath):
-        self.filepath = filepath
-        self.raw_data = RawData(filepath)
-        self.time_reference = datetime.utcfromtimestamp(
-            self.raw_data.meta_data['time_reference_seconds_since_1970'])
-        self.points = []
-        shape = self.raw_data.data.shape
-        for i in range(shape[0]):
-            for j in range(shape[1]):
-                timestamp = self.time_reference + timedelta(
-                        milliseconds=int(self.raw_data.deltatime[i]))
-                self.points.append(Point(
-                    longitude=float(self.raw_data.longitude[i, j]),
-                    latitude=float(self.raw_data.latitude[i, j]),
-                    value=float(self.raw_data.data[i, j]),
-                    quality=float(self.raw_data.quality[i, j]),
-                    timestamp=timestamp,
-                ))
+    # convert deltatime to timestamps
+    deltatime_arr = numpy.repeat(
+        deltatime, pixel_per_line).reshape(n_lines, -1)
+    deltatime_arr = deltatime_arr[mask]  # filter for missing data
+    # add (milli-)seconds since 1970
+    deltatime_arr = numpy.add(deltatime_arr, time_reference * 1000)
+    timestamps = pandas.to_datetime(deltatime_arr, utc=True, unit='ms')
 
-    def filter_by_quality(self, minimal_quality):
-        """Filter points of the Scan by quality.
+    # convert data to geodataframe
+    return geopandas.GeoDataFrame({
+        'timestamp': timestamps,
+        'quality': quality[mask],
+        'data': data[mask]
+    },
+        geometry=geopandas.points_from_xy(
+            longitude[mask],
+            latitude[mask])
+    )
 
-        :param minimal_quality: Minimal allowed quality
-        :type minimal_quality: int
-        """
-        self.points = [p for p in self.points if p.quality >= minimal_quality]
 
-    def len(self):
-        """Get number of points in Scan.
+def filter_data(dataframe, minimal_quality=0.7):
```

Comment:
Assuming that there will be more algorithms, my suggestion would be to use more specific terms like `filter_by_quality()`

Reply:
I think we could rename that now.